### PR TITLE
Deprecate extension methods for NonEmptyList

### PR DESCRIPTION
**Pull request by franciscodr**
- [x] Applicative
- [x] Apply
- [x] Bimonad
- [x] Comonad
- [x] Eq (EqK)
- [x] Foldable
- [x] Functor
- [x] Hash
- [x] Monad
- [x] Order
- [x] Semialign
- [x] Semigroup (SemigroupK)
- [x] Show
- [x] Traverse
- [x] Unzip
- [x] Zip

### CODE REVIEW

File: arrow-core-data/src/main/kotlin/arrow/core/NonEmptyList.kt [hunk 1]
```diff
@@ -206,7 +219,7 @@ class NonEmptyList<out A>(
   inline fun <B> map(f: (A) -> B): NonEmptyList<B> =
     NonEmptyList(f(head), tail.map(f))
 
-  inline fun <B> flatMap(f: (A) -> NonEmptyListOf<B>): NonEmptyList<B> =
+  fun <B> flatMap(f: (A) -> NonEmptyListOf<B>): NonEmptyList<B> =
```
Comment (nomisRev) [inline arrow-core-data/src/main/kotlin/arrow/core/NonEmptyList.kt:222]: This should stay inlined so it can ingest `suspend` functions.

File: arrow-core-data/src/main/kotlin/arrow/core/NonEmptyList.kt [hunk 2]
```diff
@@ -266,6 +279,23 @@ class NonEmptyList<out A>(
...
+      NonEmptyList(head.combine(b.head), tail.salign(SA, b.tail).toList())
+    }
+
+  fun <B> padZip(other: NonEmptyList<B>): NonEmptyList<Tuple2<A?, B?>> =
+    NonEmptyList(Tuple2(head, other.head), tail.padZip(other.tail))
+
+  fun <B> zip(other: NonEmptyList<B>): NonEmptyList<Tuple2<A, B>> =
+    NonEmptyList(Tuple2(head, other.head), tail.zip(other.tail).map { Tuple2(it.first, it.second) })
+
+  fun <B, C> zipWith(other: NonEmptyList<B>, f: Function2<A, B, C>): NonEmptyList<C> =
```
Comment (nomisRev) [inline arrow-core-data/src/main/kotlin/arrow/core/NonEmptyList.kt:296]:
```suggestion
  fun <B, C> zip(other: NonEmptyList<B>, f: (A, B) -> C): NonEmptyList<C> =
```

This is called `zip` in Kotlin Std.

File: arrow-core-data/src/main/kotlin/arrow/core/NonEmptyList.kt [hunk 3]
```diff
@@ -282,6 +312,236 @@ class NonEmptyList<out A>(
...
+      f: NonEmptyList<F>,
+      g: NonEmptyList<G>,
+      h: NonEmptyList<H>,
+      i: NonEmptyList<I>,
+      j: NonEmptyList<J>,
+      k: NonEmptyList<K>
+    ): NonEmptyList<Tuple10<B, C, D, E, F, G, H, I, J, K>> =
+      mapN(b, c, d, e, f, g, h, i, j, k) { b, c, d, e, f, g, h, i, j, k ->
+        Tuple10(b, c, d, e, f, g, h, i, j, k)
+      }
```
Comment (nomisRev) [inline arrow-core-data/src/main/kotlin/arrow/core/NonEmptyList.kt:543]:
```suggestion
```

We discussed to deprecate `tupledN` since it's just an alias for `mapN` with `::TupleN`, and thus doesn't provide a strong benefit to justify so much code / binary. This is more in line with the Kotlin Std.

Comment (nomisRev) [reply arrow-core-data/src/main/kotlin/arrow/core/NonEmptyList.kt:543]: I'm opening a PR asap to remove this from the already finished data types.
We've already deprecated `parTupledN` in Arrow Fx Coroutines as well.

File: arrow-core-data/src/main/kotlin/arrow/core/NonEmptyList.kt [hunk 4]
```diff
@@ -318,3 +578,142 @@ fun <A, G> NonEmptyListOf<Kind<G, A>>.sequence(GA: Applicative<G>): Kind<G, NonE
...
+fun <A> NonEmptyList<NonEmptyList<A>>.flatten(): NonEmptyList<A> =
+  this.flatMap(::identity)
+
+fun <A, B> NonEmptyList<Either<A, B>>.selectM(f: NonEmptyList<(A) -> B>): NonEmptyList<B> =
+  this.flatMap { it.fold({ a -> f.map { ff -> ff(a) } }, { b -> NonEmptyList.just(b) }) }
+
+fun <A, B> NonEmptyList<Tuple2<A, B>>.unzip(): Tuple2<NonEmptyList<A>, NonEmptyList<B>> =
+  this.unzipWith(::identity)
+
+fun <A, B, C> NonEmptyList<C>.unzipWith(f: Function1<C, Tuple2<A, B>>): Tuple2<NonEmptyList<A>, NonEmptyList<B>> =
```
Comment (nomisRev) [inline arrow-core-data/src/main/kotlin/arrow/core/NonEmptyList.kt:596]:
```suggestion
fun <A, B, C> NonEmptyList<C>.unzipWith(f: (C) -> Tuple2<A, B>): Tuple2<NonEmptyList<A>, NonEmptyList<B>> =
```

File: arrow-core/src/main/kotlin/arrow/core/extensions/nonemptylist/applicative/NonEmptyListApplicative.kt [hunk 5]
```diff
@@ -32,8 +32,8 @@ internal val applicative_singleton: NonEmptyListApplicative = object :
 @Deprecated(
   "@extension kinded projected functions are deprecated",
   ReplaceWith(
-  "just()",
-  "arrow.core.just"
+    "NonEmptyList.just(this)",
+    "arrow.core.NonEmptyList"
```
Comment (nomisRev) [inline arrow-core/src/main/kotlin/arrow/core/extensions/nonemptylist/applicative/NonEmptyListApplicative.kt:36]:
```suggestion
    "arrow.core.NonEmptyList", "arrow.core.just"
```

Comment (nomisRev) [reply arrow-core/src/main/kotlin/arrow/core/extensions/nonemptylist/applicative/NonEmptyListApplicative.kt:36]: Should we move towards `nonEmptyListOf` similar to the other collections constructors in Kotlin? cc\\ @raulraja

File: arrow-core/src/main/kotlin/arrow/core/extensions/nonemptylist/applicative/NonEmptyListApplicative.kt [hunk 6]
```diff
@@ -51,8 +51,8 @@ fun <A> A.just(): NonEmptyList<A> = arrow.core.NonEmptyList.applicative().run {
 @Deprecated(
   "@extension kinded projected functions are deprecated",
   ReplaceWith(
-  "unit()",
-  "arrow.core.NonEmptyList.unit"
+    "NonEmptyList.just(Unit)",
+    "arrow.core.NonEmptyList"
```
Comment (nomisRev) [inline arrow-core/src/main/kotlin/arrow/core/extensions/nonemptylist/applicative/NonEmptyListApplicative.kt:55]:
```suggestion
    "arrow.core.NonEmptyList", "arrow.core.just"
```

File: arrow-core/src/main/kotlin/arrow/core/extensions/nonemptylist/apply/NonEmptyListApply.kt [hunk 7]
```diff
@@ -89,31 +88,8 @@ fun <A, B, Z> Kind<ForNonEmptyList, A>.map2Eval(
...
-  ),
-  DeprecationLevel.WARNING
-)
-fun <A, B, Z> map(
-  arg0: Kind<ForNonEmptyList, A>,
-  arg1: Kind<ForNonEmptyList, B>,
-  arg2: Function1<Tuple2<A, B>, Z>
-): NonEmptyList<Z> = arrow.core.NonEmptyList
-   .apply()
-   .map<A, B, Z>(arg0, arg1, arg2) as arrow.core.NonEmptyList<Z>
```
Comment (nomisRev) [inline arrow-core/src/main/kotlin/arrow/core/extensions/nonemptylist/apply/NonEmptyListApply.kt:116]: Why is this removed?

File: arrow-core/src/main/kotlin/arrow/core/extensions/nonemptylist/apply/NonEmptyListApply.kt [hunk 8]
```diff
@@ -148,8 +125,8 @@ fun <A, B, Z> mapN(
 @Deprecated(
   "@extension kinded projected functions are deprecated",
   ReplaceWith(
-  "map(arg0, arg1, arg2, arg3)",
-  "arrow.core.NonEmptyList.map"
+    "NonEmptyList.tupledN(arg0, arg1, arg2).map(arg3)",
```
Comment (nomisRev) [inline arrow-core/src/main/kotlin/arrow/core/extensions/nonemptylist/apply/NonEmptyListApply.kt:128]: These should be deprecated towards `NonEmptyList.mapN` and we should deprecate `tupeldN` towards `mapN`.

```suggestion
    "NonEmptyList.mapN(arg0, arg1, arg2) { a, b, c -> arg3(Tuple3(a, b, c)) }",
```

File: arrow-core/src/main/kotlin/arrow/core/extensions/nonemptylist/foldable/NonEmptyListFoldable.kt [hunk 9]
```diff
@@ -97,8 +86,8 @@ fun <A> Kind<ForNonEmptyList, A>.fold(arg1: Monoid<A>): A = arrow.core.NonEmptyL
 @Deprecated(
   "@extension kinded projected functions are deprecated",
   ReplaceWith(
-  "reduceLeftToOption(arg1, arg2)",
-  "arrow.core.reduceLeftToOption"
+    "Option.fromNullable(this.fix().reduceOrNull(arg1, arg2))",
```
Comment (nomisRev) [inline arrow-core/src/main/kotlin/arrow/core/extensions/nonemptylist/foldable/NonEmptyListFoldable.kt:89]: 👌

File: arrow-core/src/main/kotlin/arrow/core/extensions/nonemptylist/zip/NonEmptyListZip.kt [hunk 10]
```diff
@@ -48,8 +48,8 @@ fun <A, B> Kind<ForNonEmptyList, A>.zip(arg1: Kind<ForNonEmptyList, B>): NonEmpt
 @Deprecated(
   "@extension kinded projected functions are deprecated",
   ReplaceWith(
-  "zipWith(arg1, arg2)",
-  "arrow.core.zipWith"
+  "fix().zipWith<B>(arg1.fix(), arg2)",
```
Comment (nomisRev) [inline arrow-core/src/main/kotlin/arrow/core/extensions/nonemptylist/zip/NonEmptyListZip.kt:51]:
```suggestion
  "fix().zip<B>(arg1.fix(), arg2)",
```

**Review by nomisRev — COMMENTED** [review]
I think we're missing `traverseEither`, `traverseValidated`, `sequenceEither`, `sequenceValidated` on `NonEmptyList`. If we don't add concrete variants of this it will use the ones defined on `Iterable<A>` and will result in a return type of `Either<A, List<B>>` instead of `Either<A, NonEmptyList<B>>`.

Couples nits, and we agreed to deprecate `tupledN` this still needs to be done for the already merged data types.

File: arrow-core-data/src/main/kotlin/arrow/core/NonEmptyList.kt (hunk 1 above)
Comment (franciscodr) [reply arrow-core-data/src/main/kotlin/arrow/core/NonEmptyList.kt:222]: The problem of keeping it inlined is the compiler throws an error that says the method is too long when compiling `mapN`. `flatMap` is called several times in that method so the code is copied several times too, causing the above-mentioned error.

File: arrow-core/src/main/kotlin/arrow/core/extensions/nonemptylist/apply/NonEmptyListApply.kt (hunk 7 above)
Comment (franciscodr) [reply arrow-core/src/main/kotlin/arrow/core/extensions/nonemptylist/apply/NonEmptyListApply.kt:116]: I guess I removed it accidentally. I will put it back.

File: arrow-core-data/src/main/kotlin/arrow/core/NonEmptyList.kt (hunk 1 above)
Comment (nomisRev) [reply arrow-core-data/src/main/kotlin/arrow/core/NonEmptyList.kt:222]: 🤯 That's bad because we want `mapN` to be inlined too. I've never ran into such an error actually.

cc\\ @raulraja

File: arrow-core/src/main/kotlin/arrow/core/extensions/nonemptylist/apply/NonEmptyListApply.kt (hunk 7 above)
Comment (franciscodr) [reply arrow-core/src/main/kotlin/arrow/core/extensions/nonemptylist/apply/NonEmptyListApply.kt]: Fixed, I messed up a bit that part of the code.

**Review by nomisRev — APPROVED** [review]
Great alternative for mapN 👏 👏 Looks perfect to me!

File: arrow-core-data/src/main/kotlin/arrow/core/NonEmptyList.kt (hunk 1 above)
Comment (raulraja) [reply arrow-core-data/src/main/kotlin/arrow/core/NonEmptyList.kt]: I've never ran into that error either, @franciscodr do you have an example where we can reproduce this?

**Review by raulraja — APPROVED** [review]
thanks @franciscodr !
